addons/addon-base-raas/packages/base-raas-services/lib/study/study-permission-service.js

```diff
@@ -16,361 +17,543 @@
+  // compatible with existing internal studies. Remember the accessType represents the
+  // the maximum permissions allowed on the study.
+  // Notice that when we have accessType = readwrite, we don't clear the readonly users
+  // nor the writeonly users, this is because (as mentioned above) accessType presents
+  // the maximum permissions allowed.
+  if (isReadonly(studyEntity)) {
+    // all users who had readwrite access need to be demoted to readonlyUsers
+    if (!_.isEmpty(entity.readwriteUsers)) {
+      entity.readonlyUsers = _.uniq([...entity.readonlyUsers, ...entity.readwriteUsers]);
+      entity.readwriteUsers = [];
+    }
+    entity.writeonlyUsers = [];
+  } else if (isWriteonly(studyEntity)) {
+    // all users who had readwrite access need to be demoted to writeonlyUsers
+    if (!_.isEmpty(entity.readwriteUsers)) {
+      entity.writeonlyUsers = _.uniq([...entity.writeonlyUsers, ...entity.readwriteUsers]);
+      entity.readwriteUsers = [];
+    }
+    entity.readonlyUsers = [];
+  }
+
+  return entity;
+};
+
+const toUserPermissionsEntity = (dbEntity = {}) => {
+  const entity = _.omit(dbEntity, ['recordType', 'uid', 'id']);
+  const uid = _.get(dbEntity, 'id', '').slice('User:'.length);
+  return { ...getEmptyUserPermissions(), ...entity, uid };
 };
-const permissionLevels = ['admin', 'readonly', 'writeonly', 'readwrite'];
 
+/**
+ * IMPORTANT: This is a strict service delegate. It means that it shouldn't be used or accessed
+ * directly.  Only one service (the delegator) is expected to use this service. In this case,
+ * the delegator service is the StudyService.  If you need access to any of the study entities
+ * and permissions information, you should use the StudyService.
+ *
+ * This strict service delegate is responsible of managing the StudyPermissionsEntity and the
+ * UserPermissionsEntity.
+ */
 class StudyPermissionService extends Service {
   constructor() {
     super();
-    this.dependency(['dbService', 'jsonSchemaValidationService', 'lockService']);
+    this.dependency([
+      'dbService',
+      'jsonSchemaValidationService',
+      'authorizationService',
+      'auditWriterService',
+      'lockService',
+      'userService',
+    ]);
   }
 
   async init() {
-    // Get services
-    this.jsonSchemaValidationService = await this.service('jsonSchemaValidationService');
-    this.lockService = await this.service('lockService');
-
     // Setup DB helpers
-    const dbService = await this.service('dbService');
-    this.tableName = this.settings.get(settingKeys.tableName);
-    this._getter = () => dbService.helper.getter().table(this.tableName);
-    this._updater = () => dbService.helper.updater().table(this.tableName);
-    this._query = () => dbService.helper.query().table(this.tableName);
-    this._deleter = () => dbService.helper.deleter().table(this.tableName);
-    this._scanner = () => dbService.helper.scanner().table(this.tableName);
+    const [dbService] = await this.service(['dbService']);
```

Comment:
Why do we need a list to initialize a single service? Just curious..

Reply:
It was just a left over from refactoring, there used to be more than one service.
There is no harm from leaving it the way it is, but will change it anyway.